stl/src/vector_algorithms.cpp

```diff
@@ -24,12 +24,22 @@ static void _Reverse_tail(_BidIt _First, _BidIt _Last) noexcept {
     }
 }
 
-static size_t _Byte_length(void* _First, void* _Last) noexcept {
-    return static_cast<unsigned char*>(_Last) - static_cast<unsigned char*>(_First);
+template <class _BidIt, class _OutIt>
+static _OutIt _Reverse_copy_tail(_BidIt _First, _BidIt _Last, _OutIt _Dest) noexcept {
+    for (; _First != _Last;) {
+        *_Dest++ = *--_Last;
+    }
+    return _Dest;
+}
+
+template <class _Ptr>
+static size_t _Byte_length(_Ptr _First, _Ptr _Last) noexcept {
+    return static_cast<const unsigned char*>(_Last) - static_cast<const unsigned char*>(_First);
 }
 
-static void _Advance_bytes(void*& _Target, ptrdiff_t _Offset) noexcept {
-    _Target = static_cast<unsigned char*>(_Target) + _Offset;
+template <class _Ptr>
+static void _Advance_bytes(_Ptr& _Target, ptrdiff_t _Offset) noexcept {

```

Comment:
Why did this get templated? All the callers still look like they're void*?

And why are we assigning a const char* into it?